pkg/prometheus/agent/common.go

```diff
@@ -0,0 +1,75 @@
+// Copyright 2023 The prometheus-operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheusagent
+
+import (
+	v1 "k8s.io/api/core/v1"
+
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"github.com/prometheus-operator/prometheus-operator/pkg/operator"
+	prompkg "github.com/prometheus-operator/prometheus-operator/pkg/prometheus"
+)
+
+func buildImagePath(
+	cpf monitoringv1.CommonPrometheusFields,
+	c *prompkg.Config,
+) (string, error) {
+	pImagePath, err := operator.BuildImagePath(
+		operator.StringPtrValOrDefault(cpf.Image, ""),
+		operator.StringValOrDefault("", c.PrometheusDefaultBaseImage),
+		operator.StringValOrDefault(cpf.Version, operator.DefaultPrometheusVersion),
+		"",
+		"",
+	)
+
+	if err != nil {
+		return "", err
+	}
+
+	return pImagePath, nil
+}
+
+func buildAgentArgs(
+	cpf monitoringv1.CommonPrometheusFields,
+	cg *prompkg.ConfigGenerator,
+) []monitoringv1.Argument {
+	promArgs := prompkg.BuildCommonPrometheusArgs(cpf, cg)
+	promArgs = appendAgentArgs(promArgs, cg, cpf.WALCompression)
```

Comment:
Since WALCompression exists both for Server and Agent, I'm not sure there's a reason why it can't be handled by BuildCommonPrometheusArgs()?

Reply:
Yeah I'll look further if any pieces can also be refactored into the server mode. Refactoring agent's modes is Step 1.

Reply:
WALCompression exists in both Server and Agent. But the args are different:

- Server arg: https://github.com/prometheus-operator/prometheus-operator/blob/cbf0f66c687656b6a96e2acb8b618ef0aaaf138c/pkg/prometheus/server/statefulset.go#L567
- Agent arg: https://github.com/prometheus-operator/prometheus-operator/blob/cbf0f66c687656b6a96e2acb8b618ef0aaaf138c/pkg/prometheus/agent/statefulset.go#L436

Reply:
Ah bummer! Sorry for the noise then :)